databricks_cli/sdk/service.py

```diff
@@ -519,25 +523,45 @@ def list_test(self, path, headers=None):
             _data['path'] = path
         return self.client.perform_query('GET', '/dbfs-testing/list', data=_data, headers=headers)
 
-    def put(self, path, contents=None, overwrite=None, headers=None):
+    def put(self, path, contents=None, overwrite=None, headers=None, src_path=None):
         _data = {}
+        _files = None
         if path is not None:
             _data['path'] = path
         if contents is not None:
-            _data['contents'] = contents
+            # Because terminal might add trailing newlines, they need to be encoded properly.
+            encoded_contents = base64.b64encode(contents.encode('utf-8'))
+            _data['contents'] = encoded_contents.decode("utf-8")
         if overwrite is not None:
             _data['overwrite'] = overwrite
-        return self.client.perform_query('POST', '/dbfs/put', data=_data, headers=headers)
-
-    def put_test(self, path, contents=None, overwrite=None, headers=None):
-        _data = {}
+        if src_path is not None:
+            # @self.client sets Content-Type 'text/json' by default.
+            # For multipart/form-data POST Content-Type should be set automatically
+            # to decode 'Boundary' parameter.
+            headers = {'Content-Type': None}
+            filename = os.path.basename(src_path)
+            _files = {'file': (filename, open(src_path, 'rb'), 'multipart/form-data')}
```

Comment:
What's this struct/tuple you're passing for files? Is this format defined somewhere or did you create it? How does the request know to send these files as a multipart upload?

Reply:
If you go to `perform_query`, it passes a `file=` argument to the request. When that is the case, POST requests becomes a multipart upload. It is explained in requests docs: https://docs.python-requests.org/en/master/user/quickstart/#post-a-multipart-encoded-file.